docs/rfc/rfc-19-random-beacon-v2.adoc

```diff
@@ -0,0 +1,225 @@
+maintenance pool. The reward is paid to the DKG result submitter in the transaction
+approving the DKG result. The logic triggering new group selection will be embedded 
+in relay entry request transaction and should be as cheap as possible - it only
+locks the sortition pool and emits a group creation start event.
+
+In case the DKG result has not been submitted on time, anyone can unlock the pool
+and receive a fixed, governable reward for reporting DKG timeout. The reward is
+paid from the maintenance pool.
+
+Just like in the case of relay entry submission, the order in which operators
+are supposed to submit DKG result is enforced on-chain and operators who missed
+their round are removed from the sortition pool for at least 2 weeks and do
+not earn rewards for that time. The first member eligible to submit the DKG result
+is a member with index `1`, then also `2`. For example, if member `9` submitted
+DKG result, members with indexes `{1, 2, 3, 4, 5, 6, 7, 8}` should be removed
+from the sortition pool in the same transaction in which member `9` submitted
+DKG result.
+
+The default time for each member to submit DKG result should be set to 10 blocks
+and should become a governable parameter.
+
+To make this scheme work, both the relay entry and the DKG result submission
+transactions need to have predictable gas costs. Although it is guaranteed for
+the latter, the former, in v1 of the beacon, depends on the gas cost of
+executing a callback and needs to be optimized - more on that in the
+ <<callbacks,Callbacks>> section.
+
+T rewards will be distributed continuously to all operators in the beacon
+sortition pool, just like in the case of tBTC v2.
+
+[[callbacks]]
+=== Callbacks
+
+In v1 of the random beacon, a callback is executed in the same transaction in
+which relay entry is submitted with a gas limit of 2M. This allows applications
+using the random beacon to avoid an additional complexity at their end but comes
+at the cost of random beacon operators who need to have enough ETH on their
+balance. This approach is not compatible with the idea that all rewards will be
+solely in T given that it is impossible to establish the cost of executing a
+callback in T.
+
+The fact the full - even the most complex - callback is executed in the same
+transaction in which the relay entry is submitted gives an impression of better
+security. This impression is false though, given that the entry to be submitted
+is visible in the mempool and smart attackers can have their transactions mined
+faster to, for example, put the sortition pool using the entry for group
+selection in the desired state. This issue needs to be solved on the sortition
+pool side with initiation time for new operators in the pool and/or state lock.
+
+To ensure predictable cost of relay entry submission transaction, we should
+eliminate callbacks altogether. It is up to the application to act when a new
```

Comment:
I'm not 100% sure of this, compared to enforcing a low gas budget (e.g. 100k) for the callback, but may warm up to the idea

Reply:
Good point. I removed low gas budget callbacks in the last commit making sure relay entry submit transaction is as cheap as possible, moving the responsibility/complexity entirely to the application. But now I think it could be too much for the application to handle...

Let's look at it from the perspective of tBTC v2.

With a callback: someone sees that the maintenance pool allows creating a new wallet and they submit a transaction. In this transaction, sortition pool is locked and a new relay entry is requested. Beacon produces an entry and executes a callback on the application. The callback stores a new seed and emits an event that is handled by clients performing sortition and generating group public key.

With no callback: someone sees that the maintenance pool allows creating a new wallet and they submit a transaction. In this transaction, sortition pool is locked and a new relay entry is requested. Beacon produces an entry and stores it in the beacon contract. Off-chain clients of tBTC v2 need to monitor the beacon contract to know when a new entry was submitted so that they can perform sortition and start generating group public key. This version is more complex but still possible to handle. The real complexity though begins when someone wants to challenge the submitted DKG result: we'd need to make sure the same relay entry is used for resolving the challenge as the one that was used by clients to perform sortition. 

It seems to me that implementing a low gas budget callback is much simpler than handling synchronization of entry on the application side.

What I'd propose to do:
- non-reimbursable, low gas budget callback,
- initial limit set to 50k gas or less: should be enough to SSTORE entry and emit an event,
- gas limit for the callback will be a governable value,
- for simplicity, we'd require that the contract being called has a function with a specific signature so that the only callback parameter is the contract address.

What do you think?

Reply:
Please see https://github.com/keep-network/keep-core/pull/2585/commits/c719de341690bfeb6c20cf0e47b566157a37b8ca.